internal/pkg/service/stream/storage/node/readernode/sliceupload/operator.go

```diff
@@ -0,0 +1,363 @@
+		clock:      d.Clock(),
+		logger:     d.Logger().WithComponent("storage.node.operator.slice.upload"),
+		publicAPI:  d.KeboolaPublicAPI(),
+		volumes:    d.Volumes(),
+		statistics: d.StatisticsL1Cache(),
+		storage:    d.StorageRepository(),
+		plugins:    d.Plugins(),
+	}
+
+	// Graceful shutdown
+	wg := &sync.WaitGroup{}
+	ctx, cancel := context.WithCancel(context.Background())
+	d.Process().OnShutdown(func(_ context.Context) {
+		o.logger.Info(ctx, "closing slice upload operator")
+
+		// Stop mirroring
+		cancel()
+		wg.Wait()
+		ctx := context.WithoutCancel(ctx)
+		for {
+			remainingToUpload := 0
+			o.slices.ForEach(func(_ model.SliceKey, data *sliceData) (stop bool) {
+				wg.Add(1)
+				// Prevent multiple checks of the same slice
+				if !data.Lock.TryLock() {
+					return false
+				}
+				defer func() {
+					data.Lock.Unlock()
+					wg.Done()
+				}()
+
+				// Slice uploading is already in progress
+				if data.Uploading {
+					remainingToUpload++
+					return false
+				}
+
+				// Slice has been modified by some previous check, but we haven't received an updated version from etcd yet
+				if data.Processed {
+					return false
+				}
+
+				if !data.Slice.Retryable.Allowed(o.clock.Now()) {
+					remainingToUpload++
+					return false
+				}
+
+				volume, err := o.volumes.Collection().Volume(data.Slice.SliceKey.VolumeID)
+				if err != nil {
+					o.logger.Errorf(ctx, "unable to upload slice. No volume found for key: %v", data.Slice.SliceKey.VolumeID)
+					return true
+				}
+
+				switch data.Slice.State {
+				case model.SliceUploading:
+					remainingToUpload++
+					o.uploadSlice(ctx, volume, data)
+				default:
+					// nop
+				}
+				return false
+			})
+			if remainingToUpload == 0 {
+				break
+			}
+		}
+		o.logger.Info(ctx, "closed slice upload operator")
+	})
+
+	// Mirror uploading slices
+	{
+		o.slices = etcdop.SetupMirrorMap[model.Slice, model.SliceKey, *sliceData](
+			d.StorageRepository().Slice().GetAllInLevelAndWatch(ctx, model.LevelLocal, etcd.WithPrevKV()),
+			func(_ string, slice model.Slice) model.SliceKey {
+				return slice.SliceKey
+			},
+			func(_ string, slice model.Slice, rawValue *op.KeyValue, oldValue **sliceData) *sliceData {
+				out := &sliceData{
+					Slice: &slice,
+				}
+
+				// Keep the same lock, to prevent parallel processing of the same slice.
+				// No modification from another code is expected, but just to be sure.
+				if oldValue != nil {
+					out.Lock = (*oldValue).Lock
+				} else {
+					out.Lock = &sync.Mutex{}
+				}
+
+				return out
+			},
+		).
+			WithFilter(func(event etcdop.WatchEvent[model.Slice]) bool {
+				return o.volumes.Collection().HasVolume(event.Value.VolumeID)
+			}).
+			BuildMirror()
+		if err = <-o.slices.StartMirroring(ctx, wg, o.logger); err != nil {
+			return err
+		}
+	}
+
+	// Mirror already uploaded slices for updating manifest
+	{
+		o.uploadedSlices = etcdop.SetupMirrorMap[model.Slice, model.SliceKey, *uploadedSliceData](
```

Comment:
Only uploaded slices are neede to be mirrored in order to update manifest of single `FileKey`.

Reply:
I wouldn't use a mirror here, since you don't need that data often, you can easily load it from the repository.

Another situation is when ALL slices in Uploading status are checked every second - and their retry conditions. There can be 100-10000 records there.

Reply:
Removed. Repository used instead and used in keboolabridge as it is specific for keboola storage provider.